sycl/doc/extensions/proposed/sycl_ext_oneapi_raw_kernel.asciidoc

```diff
@@ -0,0 +1,563 @@
+  using value_t = property_value<range_kernel_key, Dims, Name>;
+};
+
+template<int Dims, typename Name = void>
+inline constexpr range_kernel_key::value_t<Dims, Name> range_kernel;
+
+template<>
+struct is_property_key<range_kernel_key> : std::true_type {};
+----
+!====
+
+Indicates that the function is a raw kernel that is invoked with a simple
+`range` iteration space of `Dims` dimensions.
+When the `Name` parameter is not `void`, it is a type that represents the name
+of the kernel.
+
+The `property_value` struct has the following member variables and type
+aliases:
+
+[%header,cols="1,1"]
+!====
+!Member
+!Description
+
+a!
+[source]
+----
+static constexpr int dimensions = Dims
+----
+!
+The number of dimensions of the kernel's range.
+
+a!
+[source]
+----
+using name = Name
+----
+!
+The type-name of the kernel.
+!====
+
+a|
+*`nd_range_kernel` property*
+
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+struct nd_range_kernel_key {
+  template <int Dims, typename Name>
+  using value_t = property_value<nd_range_kernel_key, Dims, Name>;
+};
+
+template<int Dims, typename Name = void>
+inline constexpr nd_range_kernel_key::value_t<Dims, Name> nd_range_kernel;
+
+template<>
+struct is_property_key<nd_range_kernel_key> : std::true_type {};
+----
+!====
+
+Indicates that the function is a raw kernel that is invoked with an
+`nd_range` iteration space of `Dims` dimensions.
+When the `Name` parameter is not `void`, it is a type that represents the name
+of the kernel.
+
+The `property_value` struct has the following member variables and type
+aliases:
+
+[%header,cols="1,1"]
+!====
+!Member
+!Description
+
+a!
+[source]
+----
+static constexpr int dimensions = Dims
+----
+!
+The number of dimensions of the kernel's range.
+
+a!
+[source]
+----
+using name = Name
+----
+!
+The type-name of the kernel.
+!====
+
+a|
+*`single_task_kernel` property*
+
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+struct single_task_kernel_key {
+  template <typename Name>
+  using value_t = property_value<single_task_kernel_key, Name>;
+};
+
+template<typename Name = void>
+inline constexpr single_task_kernel_key::value_t<Name> single_task_kernel;
+
+template<>
+struct is_property_key<single_task_kernel_key> : std::true_type {};
+----
+!====
+
+Indicates that the function is a raw kernel that is invoked via `single_task`
+(i.e. without any iteration space).
+When the `Name` parameter is not `void`, it is a type that represents the name
+of the kernel.
+
+The `property_value` struct has the following member type alias:
+
+[%header,cols="1,1"]
+!====
+!Member
+!Description
+
+a!
+[source]
+----
+using name = Name
+----
+!
+The type-name of the kernel.
+!====
+|====
+
+When a function is defined as a raw kernel, each parameter to the function is
+a kernel argument, which must abide by the rules for allowable kernel parameter
+types specified in section 4.12.4 "Rules for parameter passing to kernels" of
+the core SYCL specification.
```

Comment:
I think that we should be more explicit here, because raw functions instead of lambdas provide more possibilities for users and therefore create more corner cases to handle:

What if user specifies default values for some arguments? Should it trigger an error, or it would still be allowed, but won't mean anything? Maybe they will be pre-recorded to `set_arg` can be omitted for those arguments?

What if I want to have `kernel_handler` in a kernel? What if I pass it in between other arguments? So far the only functionality it provides is specialization constants, but there could be more of it going forward (either in core SYCL or in some extensions). I understand that we may not be ready to provide such functionality right away (or even ever), but we probably need to an explicit note about it here.

What if user defined a kernel as a variadic function?

With kernels defined as functions, it unlocks possibilities to call one kernel from another. Should that be legal and if so, how should that behave? Note: OpenCL C prohibits calling a kernel from another kernel if I'm not mistaken. At SPIR-V level `OpEntryPoint` (kernel) cannot be a target of an `OpCall` instruction, but I think that we have solved that by duplicating such entry points and turning a copy into a regular `OpFunction`.

In [4.12.1. Defining kernels as named function objects](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:interfaces.kernels.as.function-objects), we have the following paragraph:

> The `operator()` member function must be const-qualified, and it may take different parameters depending on the data accesses defined for the specific kernel. If the `operator()` function writes to any of the member variables, the behavior is undefined.

We have a different situations here and there is no questions like "what if happens I modified my functor?". However, there could still be questions like: are those arguments shared within a work-group, or each work-item receives its own copy? Can we write to those arguments within kernels?

Reply:
Oh, one more thing to add, but it is likely more about implementation design than the spec: as I understand it, nothing prohibits user from calling raw kernel as a regular host function. We need to make sure that it is not broken by some of our implementation details

Reply:
Thanks for these comments, @AlexeySachkov.  I agree that some are issues, but others are not:

> What if user specifies default values for some arguments? Should it trigger an error, or it would still be allowed, but won't mean anything? Maybe they will be pre-recorded to set_arg can be omitted for those arguments?

This is a good point that I had not thought of.  I captured this as an open issue (397668c47c09a0b0a6cbe82d05d1601e6ded8187), which needs to get resolved.

> What if I want to have kernel_handler in a kernel?

I'm not worried about this.  If we need to use the `kernel_handler` in one of these kernels, I think we should add another free function to [sycl_ext_oneapi_free_function_queries](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_queries.asciidoc) which retrieves it.  Or, better yet, we can define a new way to get a specialization constant that does not require the `kenel_handler`.

> What if user defined a kernel as a variadic function?

We already prohibit device functions from being variadic, so I don't think this is an issue.  See [section 5.4](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:language.restrictions.kernels) of the SYCL spec.

> With kernels defined as functions, it unlocks possibilities to call one kernel from another.

I think this is not a problem because the device compiler creates a wrapper around the user's kernel function, and it's this wrapper that is the "true" kernel (i.e. the **OpEntryPoint**).  Since the wrapper isn't visible to the user, the user can't call it.

> However, there could still be questions like: are those arguments shared within a work-group, or each work-item receives its own copy? Can we write to those arguments within kernels?

I clarified this in 817bff3216f8bc742a7acc96a4fcbe8c68910027.

Reply:
> > What if user defined a kernel as a variadic function?
> 
> We already prohibit device functions from being variadic, so I don't think this is an issue. See [section 5.4](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:language.restrictions.kernels) of the SYCL spec.

Strictly speaking, the spec prohibits calls to variadic functions from device code, but not definition of device variadic functions. I assume that the latter is implied, but still, some weird dummy corner cases are possible.

> No variadic functions are allowed to be called in a [device function](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#device-function).

---

> > What if I want to have `kernel_handler` in a kernel?
>
> I'm not worried about this. If we need to use the `kernel_handler` in one of these kernels, I think we should add another free function to [sycl_ext_oneapi_free_function_queries](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_queries.asciidoc) which retrieves it. Or, better yet, we can define a new way to get a specialization constant that does not require the `kenel_handler`.

My main point here is that we may as well explicitly prohibit having `kernel_handler` as an argument of a free function kernel to simplify our implementation (by dismissing this case early on). Otherwise, it is legal to write the following (albeit slightly useless code) and we will have to deal with it (not sure though if it would necessary be complex at all since we already have support for `kernel_handler` for regular kernels):

```c++
SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
void iota(float start, kernel_handler kh, float *ptr) {
  // Note that I can place `kernel_handler` argument in a random position
  // ... some code here ...
}

q.parallel_for(range{1024}, [=](item<1> it, kernel_handler kh) {
  iota(42, kh, usmPtr);
});
```

Reply:
> Strictly speaking, the spec prohibits calls to variadic functions from device code, but not definition of device variadic functions. I assume that the latter is implied, but still, some weird dummy corner cases are possible.

OK, fair point.  I can explicitly add a restriction against variadic free function kernels.

> My main point here is that we may as well explicitly prohibit having `kernel_handler` as an argument of a free function kernel to simplify our implementation (by dismissing this case early on).

I agree that we should prohibit this, so I think it's only a mater of wording in the spec.  Personally, I think the spec already prohibits this because it says that the function parameters must abide by the restrictions for kernel arguments defined in the core SYCL specification [section 4.12.4](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:kernel.parameter.passing).  Those restrictions already prevent passing a kernel argument of type `kernel_handler`.

I think `kernel_handler` is not really a kernel argument using the core SYCL specification's definition.  Instead, it's like the `reducer` class ... a special thing that the SYCL implementation can pass to the kernel.  It's not something the user can pass as a kernel parameter as a normal variable.

I'll try to think of better wording for this in the extension spec.

BTW, the code snippet you show above seems pretty weird.  You define `iota` as a free function kernel, but you don't invoke it as a free function kernel.  Instead, you just call it as a regular device function from a different kernel.  Were you intentionally trying to illustrate that case?

Reply:
I clarified that `kernel_handler` is not allowed as an argument type in 83193df7236bf59a9d76f19cf785f723022ea196.

I clarified that variadic arguments and default parameter values are not allowed in 326b502ed5c5fffbc949a16f54fff176818bd2ec.